src/main/auth/o-auth2.ts

```diff
@@ -0,0 +1,149 @@
+import assert from 'assert';
+import { URLSearchParams } from 'url';
+import { Request } from '../request';
+import { RequestAuthorization, AuthRetryConfig } from '../definitions';
+
+export enum OAuth2GrantType {
+    CLIENT_CREDENTIALS = 'client_credentials',
+    REFRESH_TOKEN = 'refresh_token',
+}
+
+export interface OAuth2Params {
+    // required if we want to enforce refreshing..
+    clientId: string;
+    tokenUrl: string;
+
+    // optional
+    clientSecret?: string;
+    refreshToken?: string;
+    accessToken?: string;
+    expiresAt?: number;
+}
+
+export class OAuth2 implements RequestAuthorization, OAuth2Params {
+    retryConfig: AuthRetryConfig;
+
+    clientId: string;
+    tokenUrl: string;
+    clientSecret?: string;
+    refreshToken?: string;
+    accessToken?: string;
+    expiresAt?: number;
+
+    constructor(params: OAuth2Params, retryConfig?: AuthRetryConfig) {
+        this.tokenUrl = params.tokenUrl;
+        this.clientId = params.clientId;
+        this.refreshToken = params.refreshToken;
+        this.accessToken = params.accessToken;
+        this.expiresAt = params.expiresAt;
+        this.clientSecret = params.clientSecret;
+
+        const defaultRetryConfig = {
+            delay: 1000,
+            attempts: 5,
+            statusCodesToRetry: [[401, 401]]
+        };
+
+        this.retryConfig = retryConfig || defaultRetryConfig;
+    }
+
+    async getHeader(): Promise<string> {
+        /** if access token available,
+         *  -> check expiresAt, if not expired -> return it
+         *
+         * if refreshToken available
+         *  -> try refreshing, if succeed, update accessToken, refreshToken and expiresAt
+         *
+         * if above fails or refreshToken is not available,
+         *  -> check grant_type is client_credentials if it is, get the tokens
+         *
+         * note: do not handle authorization_code grant type
+         */
+
+        if (this.accessToken/*  && this.expiresAt && Date.now() < this.expiresAt */) {
```

Comment:
Wondering why these are disabled 🤔

Reply:
Oh yeah, I meant to ask your opinion on it. I thought about the scenario when a user just simply wants to instantiate OAuth2 with accessToken, expiresAt should be provided along with accessToken otherwise it will try to refresh tokens. Or should we ignore this case and advise them to use just Bearer? 🤔

Reply:
Depends. If client only cares about sending a specific access token and doesn't support refreshing it, then it makes little sense to use OAuth flow b/c it wouldn't have clientId, clientSecret, etc. So Bearer looks fine in this case.